pkg/gofr/datasource/openai/client.go

```diff
@@ -0,0 +1,196 @@
+package openai
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/google/go-querystring/query"
+	"go.opentelemetry.io/otel/trace"
+)
+
+type Config struct {
+	APIKey       string
+	Model        string
+	BaseURL      string
+	Timeout      time.Duration
+	MaxIdleConns int
+}
+
+type Client struct {
+	config     *Config
+	logger     Logger
+	metrics    Metrics
+	tracer     trace.Tracer
+	httpClient *http.Client
+}
+
+var (
+	ErrorMissingAPIKey = errors.New("api key not provided")
+)
+
+type ClientOption func(*Client)
+
+func WithClientHTTP(httpClient *http.Client) func(*Client) {
+	return func(c *Client) {
+		c.httpClient = httpClient
+	}
+}
+
+func WithClientTimeout(d time.Duration) func(*Client) {
+	return func(c *Client) {
+		c.httpClient.Timeout = d
+	}
+}
+
+func NewClient(config *Config, opts ...ClientOption) (*Client, error) {
+	if config.APIKey == "" {
+		return nil, ErrorMissingAPIKey
+	}
+
+	if config.BaseURL == "" {
+		config.BaseURL = "https://api.openai.com"
+	}
+
+	// Use the provided HTTP client or create a new one with defaults
+	c := &Client{
+		config: config,
+		httpClient: &http.Client{
+			Timeout: config.Timeout,
+			Transport: &http.Transport{
+				MaxIdleConns:    config.MaxIdleConns,
+				IdleConnTimeout: 30 * time.Second,
+			},
+		},
+	}
+
+	for _, opt := range opts {
+		opt(c)
+	}
+
+	return c, nil
+}
+
+func (c *Client) UseLogger(logger interface{}) {
+	if l, ok := logger.(Logger); ok {
+		c.logger = l
+	}
+}
+
+func (c *Client) UseMetrics(metrics interface{}) {
+	if m, ok := metrics.(Metrics); ok {
+		c.metrics = m
+	}
+}
+
+func (c *Client) UseTracer(tracer any) {
+	if tracer, ok := tracer.(trace.Tracer); ok {
+		c.tracer = tracer
+	}
+}
+
+func (c *Client) InitMetrics() {
+	openaiHistogramBuckets := []float64{.05, .075, .1, .125, .15, .2, .3, .5, .75, 1, 2, 3, 4, 5, 7.5, 10}
+
+	c.metrics.NewHistogram(
+		"openai_api_request_duration",
+		"Duration of OpenAPI requests in seconds",
```

Comment:
Other are in lowercase except brands and acronyms

```suggestion
		"duration of OpenAPI requests in seconds",
```